### ISSUE TITLE
Mutex-free lookups of chain status

### ISSUE THREAD

**Issue by holiman** [ethereum/go-ethereum#15819]
It would be great to have these sync-free methods. Whenever we write to e.g. currentblock, we could set 
atomic integers to expose these methods for callers. These are used in the p2p-communication aswell as in rpc-calls, e.g. 
estimate gas. 

I implemented one of them a while ago in https://github.com/ethereum/go-ethereum/pull/15104 , but I think it has bitrotted away by now. 

```golang

	// returns current total difficulty, the same that would be returned by 
	// pm.blockchain.CurrentBlock()
	// pm.blockchain.GetTd(currentBlock.Hash(), currentBlock.NumberU64())
	// i.e , the currentBlock.
 
	td := pm.blockchain.CurrentTd()

	// Returns the number of the pm.blockchain.CurrentBlock()
	number := pm.blockchain.CurrentNumber()

	// Returns the number of pm.blockchain.CurrentFastBlock()
	fastNumber := pm.blockchain.CurrentFastNumber()


	// Returns the GasLimit() for CurrentBlock()
	gasLimit := pm.blockchain.CurrentGasLimit()
```
I believe this is the cause for https://github.com/ethereum/go-ethereum/issues/15739 , aswell as peers being dropped/unresponsive during high load -- particularly when the drive IO is the bottleneck, since the blockchain mutex will be held almost constantly.

**Comment by ryanschneider** [comment]
I haven't gotten to testing it yet, but FWIW I've cherry-picked your original commit from #15104 onto master here:

https://github.com/ryanschneider/go-ethereum/commit/4078a38f36449e07948170e2330a66cc4aa4d3b4

I'm hoping to try it out on a node that's been having trouble syncing soon.

**Comment by holiman** [comment]
Thanks @ryanschneider ! I haven't been having these problems myself since I swapped out my old (heavily degraded) SSD, so I'm having some problems verifying this myself. I have seen in various reports/stack traces that a huge number of threads have been stuck waiting on that mutex.

**Comment by holiman** [comment]
However, there's still one point where it may lock up, https://github.com/ryanschneider/go-ethereum/commit/4078a38f36449e07948170e2330a66cc4aa4d3b4#diff-75028a2cddf4c8ebf4133ad28d81ef7eR200 , where it would be nice to have a `CurrentNumber` aswell

**Comment by ryanschneider** [comment]
Good point, I'll try implementing `CurrentNumber` time permitting early next week.

**Comment by ryanschneider** [comment]
Actually implemented it today:

https://github.com/ryanschneider/go-ethereum/commit/1ef1a74a5eb6503f791bab57707c69c1abde5310

And I have it running on two nodes, both were someone out of sync.  The first one just caught up so that's a good sign, the second is almost 100,000 blocks behind and an archive node, so we'll see if it catches up any time soon.

**Comment by holiman** [comment]
Great, very interesting!

**Comment by ryanschneider** [comment]
While the patches seemed to work fine, it didn't seem to fix my exact issue (which seems to be more related to #15824).  I'll submit a new PR though w/ my changes, referencing this issue, hopefully some time today.
